### ISSUE TITLE
sled-agent: advertise underlay prefix at startup

### ISSUE THREAD

**Issue by rcgoodfellow** [oxidecomputer/omicron#3532]
We recently had an unexpected sled reboot. When the node came back up, its underlay prefix was not advertised through `ddmd`, and thus the node was not able to reach critical services such as DNS. When the sled agent is starting up, the underlay prefix should always be advertised (this is idempotent, so if it's already advertised, there is no harm).

**Comment by rmustacc** [comment]
Strictly speaking if the sled reboots it will not know what the underlay prefix to use is. It must get this via the LRTQ. So this should be part of the normal transition from being on bootstrap only to unlocking.

**Comment by rcgoodfellow** [comment]
The advertisement should go here.

- https://github.com/oxidecomputer/omicron/blob/5184ab929ce855ba9f92c90c868a7ed570e878c7/sled-agent/src/bootstrap/agent.rs#L510

**Comment by rcgoodfellow** [comment]
I went to go fix this and the issue did not come up on a basic repro attempt.

I brought up a control plane, and from ddm's perspective, things looked like this.

```
ry@masaka:~/src/omicron$ /opt/oxide/mg-ddm/ddmadm get-originated
Prefix
fd00:1122:3344:2::/64
fd00:1122:3344:3::/64
fd00:1122:3344:101::/64
fd00:1122:3344:1::/64
fdb0:a85e:45cd:bd94::/64
```

Then I simulated a reboot.

```
svcadm disable mg-ddm
svcadmvdisable sled-agent
svcadm enable mg-ddm
svcadm enable sled-agent
```

When ddm came back up it had no prefixes (expected). Then sled agent came back up, cleared out all the zones and started rebuilding them one by one. It also started putting prefixes back one by one. By the time sled agent had re-launched all the zones, all the advertisements were also back.

```
ry@masaka:~/src/omicron$ /opt/oxide/mg-ddm/ddmadm get-originated
Prefix
fd00:1122:3344:1::/64
fd00:1122:3344:2::/64
fd00:1122:3344:101::/64
fd00:1122:3344:3::/64
fdb0:a85e:45cd:bd94::/64
```

@smklein is it possible this has been fixed in recent cold boot work since this issue was opened last week?

**Comment by rcgoodfellow** [comment]
This appears to have happened again on Rack 2.

[This line](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/bootstrap/agent.rs#L687) shows up in the log, but we're not making it all the way [here](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/bootstrap/agent.rs#L816) and we don't see [this](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/bootstrap/agent.rs#L822) in the log, so we should be in the SledAgentState::Before(hardware_monitor) path.

We appear to be making it to SledServer::start so that means were not stuck at the state lock at the top of request_sled_agent.

```
BRM42220014 # cat sa.log | grep -i 'setting up sled agent server' | looker
00:01:10.893Z INFO SledAgent: setting up sled agent server
```

I think we're stuck [here](https://github.com/oxidecomputer/omicron/blob/5184ab929ce855ba9f92c90c868a7ed570e878c7/sled-agent/src/sled_agent.rs#L334C31-L334C31) as this is a loop that does not terminate until success.
If NTP depends on underlay addressing being fully advertised, i'm confused how this works in the initial bringup case.

I think moving [self.ddmd_client.advertise_prefix(request.subnet)](https://github.com/oxidecomputer/omicron/blob/5184ab929ce855ba9f92c90c868a7ed570e878c7/sled-agent/src/bootstrap/agent.rs#L630) before [SledServer::start](https://github.com/oxidecomputer/omicron/blob/5184ab929ce855ba9f92c90c868a7ed570e878c7/sled-agent/src/bootstrap/agent.rs#L587-L593) may fix this.

But I'm a bit confused about how this all works in the RSS case if NTP depends on the underlay being fully advertised, which depends on Nexus being up, which depends on NTP.

**Comment by luqmana** [comment]
> I think we're stuck [here](https://github.com/oxidecomputer/omicron/blob/5184ab929ce855ba9f92c90c868a7ed570e878c7/sled-agent/src/sled_agent.rs#L334C31-L334C31) as this is a loop that does not terminate until success.

Hmm, that shouldn't be a blocking loop there? `notify_nexus_about_self` just pushes a future that loops trying to contact nexus onto a queue that should be executed in the background.

**Comment by rcgoodfellow** [comment]
Yeah, I'm having a bit of trouble wading through all the asynchrony. Observation shows that we're getting stuck somewhere in `SledServer::start`.

**Comment by rcgoodfellow** [comment]
But I think regardless, the underlay advertisement should be pushed up before the `SledServer::start`.

**Comment by luqmana** [comment]
(Log at `/net/catacomb/data/staff/dogfood/jul-24/omicron-3753/oxide-sled-agent:default.log.0` for reference)

**Comment by luqmana** [comment]
Ah ok, so the chain of events is:
1. Bootstrap agent attempted to [start sled-agent](https://github.com/oxidecomputer/omicron/blob/7a4d7a79aef4aa16bb11227015c84d1735745beb/sled-agent/src/bootstrap/agent.rs#L560) before [advertising the sled prefix](https://github.com/oxidecomputer/omicron/blob/7a4d7a79aef4aa16bb11227015c84d1735745beb/sled-agent/src/bootstrap/agent.rs#L603).
2. Sled Agent went about starting up, eventually trying to [start the services](https://github.com/oxidecomputer/omicron/blob/7a4d7a79aef4aa16bb11227015c84d1735745beb/sled-agent/src/sled_agent.rs#L375) previously written to the ledger.
3. That means [starting the Internal DNS & NTP](https://github.com/oxidecomputer/omicron/blob/ab92330f354b1230b43ad7d440a29f6c94e9cc64/sled-agent/src/services.rs#L553-L571) services first and [waiting for timesync](https://github.com/oxidecomputer/omicron/blob/ab92330f354b1230b43ad7d440a29f6c94e9cc64/sled-agent/src/services.rs#L575) before [proceeding with all the other services](https://github.com/oxidecomputer/omicron/blob/ab92330f354b1230b43ad7d440a29f6c94e9cc64/sled-agent/src/services.rs#L610-L622).
4. But from the log we never see it get past the time-sync wait. That tracks given from the NTP log we saw that it was failing on DNS lookups to the internal DNS servers. As mentioned above, that was determined to be because of the missing underlay prefix advertisement. That is, the same advertisement we're now blocking on advertising back in step (1).

That explains why the fix in #3756 works: we're no longer blocked on timesync before trying to advertise the prefix (which in turn allows the time sync to actually complete).

But what about during RSS? Well, it similarly goes through (1) & (2) above, except that the `load_services` call [bails out pretty early](https://github.com/oxidecomputer/omicron/blob/ab92330f354b1230b43ad7d440a29f6c94e9cc64/sled-agent/src/services.rs#L544) given there's no set of services to start find in the ledger given this is during rack init. Hence we hit the "advertise prefix" path and sled agent comes up waiting for RSS to kick off.
